File: chapters/01_gateway_event_loop/v3_event_lifecycle.py
from __future__ import annotations


def _intent_from_text(text: str) -> str:
    lowered = text.lower()
    if "create" in lowered or "创建" in text:
        return "create_file"
    if "read" in lowered or "读取" in text:
        return "read_file"
    return "reply"


def lifecycle_for_message(text: str, channel: str = "cli", user: str = "u1") -> list[dict[str, str]]:
    intent = _intent_from_text(text)
    return [
        {"type": "lifecycle_start", "channel": channel, "user": user},
        {"type": "ingress", "text": text},
        {"type": "act", "intent": intent},
        {"type": "egress", "message": f"done:{intent}"},
        {"type": "lifecycle_end", "status": "ok"},
    ]


def run_gateway_loop(inputs: list[str], channel: str = "cli", user: str = "u1") -> list[dict[str, object]]:
    outputs: list[dict[str, object]] = []
    for text in inputs:
        events = lifecycle_for_message(text, channel=channel, user=user)
        outputs.append({"message": text, "events": events, "final": events[-2]["message"]})
    return outputs


def main() -> None:
    out = run_gateway_loop(["create a file", "read file"])
    print(out)


if __name__ == "__main__":
    main()

File: chapters/01_gateway_event_loop/v3_event_lifecycle.ts
// TypeScript mirror for Chapter 01: Gateway Event Loop (v3)
// Run with: npx ts-node v3_event_lifecycle.ts

interface LifecycleEvent {
  type: string;
  [key: string]: string | undefined;
}

interface MessageResult {
  message: string;
  events: LifecycleEvent[];
  final: string;
}

function intentFromText(text: string): string {
  const lowered = text.toLowerCase();
  if (lowered.includes("create") || text.includes("创建")) {
    return "create_file";
  }
  if (lowered.includes("read") || text.includes("读取")) {
    return "read_file";
  }
  return "reply";
}

function lifecycleForMessage(
  text: string,
  channel: string = "cli",
  user: string = "u1"
): LifecycleEvent[] {
  const intent = intentFromText(text);
  return [
    { type: "lifecycle_start", channel, user },
    { type: "ingress", text },
    { type: "act", intent },
    { type: "egress", message: `done:${intent}` },
    { type: "lifecycle_end", status: "ok" },
  ];
}

function runGatewayLoop(
  inputs: string[],
  channel: string = "cli",
  user: string = "u1"
): MessageResult[] {
  return inputs.map((text) => {
    const events = lifecycleForMessage(text, channel, user);
    return {
      message: text,
      events,
      final: events[events.length - 2].message ?? "",
    };
  });
}

function main(): void {
  const out = runGatewayLoop(["create a file", "read file"]);
  console.log(out);
}

main();
